File: 0001-0500/0458.Poor Pigs.cpp
/**
 *Basic idea:
 * Arrange buckets into a k-dimensional square matrix. the size of the matrix is n-by-n-by-n-by-n-...-by-n .
 * n is the number of (unposioned) buckets a pig can test in the given period.
 *Result:
 * 6 / 6 test cases passed.
 * Status: Accepted
 * Runtime: 0 ms
 * Your runtime beats 10.98% of cpp submissions.
 *Date:
 * 11/13/2016
 */
class Solution {
public:
    int poorPigs(int buckets, int minutesToDie, int minutesToTest) {
        return ceil(log(buckets)/log(minutesToTest/minutesToDie+1));
    }
};

File: 0001-0500/0458.Poor Pigs.java
/**
 First (incorrect) thought: 
 When there are p pigs, group buckets into (p+1) groups. let each pig drink bucket water of its group.
 This algorithm requires more pigs than necessary. So in sound round, there must be some bucket drunk by more than one pigs.
 How to arrange multiple pigs to drink certain shared buckets in an manageable way? We need to encode buckets!

 Encoding buckets:
 How to encode buckets? Use k-base integer! Each pig is in charge of its own k-base bit. In round i (0-indexed), each pig checks the bit that it owns. If the number of that bit equals to i, the pig drinks it. If the pig dies, then we know the posionous bucket has value i at this bit.
 Pigs can drink r rounds, from 0 to r - 1. If after the r rounds the pig is still alive, we know the posionous bucket has value r at this bit.
 So we need the minimum number of bits to encode all buckets. The number of bits is the number of pigs.

 Why is this the optimal solution?
 We detect posionous bucket by observing pigs' status. There are `buckets` number of states. With the limited round of drinks, we need to make sure pigs' status space is big enough to represent all possible states of buckets. With fewer pigs, the pig space will not be big enough to distinguish certain bucket states.
 */
class Solution {
    public int poorPigs(int buckets, int minutesToDie, int minutesToTest) {
        final int round = minutesToTest / minutesToDie;
        if (buckets != 0) {
            buckets -= 1;
        }
        int res = 0;
        while (buckets != 0) {
            buckets /= round + 1;
            res ++;
        }
        return res;
    }
}

